jans-cedarling/cedarling/src/common/cedar_schema/cedar_json/action.rs

```diff
@@ -5,27 +5,180 @@
+    /// `id_mapping` and `value_mapping`.
+    ///
+    /// The `id_mapping` param is a A `HashMap` that maps context attribute keys
+    /// (like `"access_token"`) to their corresponding `id`s (like `"acs-tkn-1"`).
+    ///
+    /// # Usage Example
+    ///
+    /// ```rs
+    /// let id_mapping = HashMap::from([("access_token".to_string(), "acs-tkn-1".to_string())]);
+    /// let json = action.build_ctx_entities_json(id_mapping, value_mapping);
+    /// ```
+    pub fn build_ctx_entity_refs_json(
+        &self,
+        mut id_mapping: HashMap<String, String>,
+    ) -> Result<Value, BuildJsonCtxError> {
+        let mut json = json!({});
+
+        for attr in self.context_entities.iter() {
+            if let CedarType::TypeName(type_name) = &attr.kind {
+                let id = match id_mapping.remove(&attr.key) {
+                    Some(val) => val,
+                    None => Err(BuildJsonCtxError::MissingIdMapping(attr.key.clone()))?,
+                };
+                let type_name = format!("{}::{}", attr.namespace, type_name);
+                json[attr.key.clone()] = json!({"type": type_name, "id": id});
+            }
+        }
+
+        Ok(json)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BuildJsonCtxError {
+    /// If an entity reference is provided but the ID is missing from `id_mapping`.
+    #[error("An entity reference for `{0}` is required by the schema but an ID was not provided via the `id_mapping`")]
+    MissingIdMapping(String),
+    /// If a non-entity attribute is provided but the value is missing from `value_mapping`.
+    #[error("A non-entity attribute for `{0}` is required by the schema but a value was not provided via the `value_mapping`")]
+    MissingValueMapping(String),
+}
+
+impl CedarSchemaJson {
+    /// Find the action in the schema
+    pub fn find_action(
+        &self,
+        action_name: &str,
+        namespace: &str,
+    ) -> Result<Option<Action>, FindActionError> {
+        let schema_entities = match self.namespace.get(namespace) {
+            Some(entities) => entities,
+            None => return Ok(None),
+        };
+
+        let action_schema = match schema_entities.actions.get(action_name) {
+            Some(schema) => schema,
+            None => return Ok(None),
+        };
+
+        let mut principal_entities = HashSet::new();
+        for res_type in action_schema.principal_types.iter().map(|s| s.as_str()) {
+            principal_entities.insert(format!("{}::{}", namespace, res_type));
+        }
+        let mut resource_entities = HashSet::new();
+        for res_type in action_schema.resource_types.iter().map(|s| s.as_str()) {
+            resource_entities.insert(format!("{}::{}", namespace, res_type));
+        }
+        let mut context_entities = HashSet::new();
+        if let Some(ctx) = &action_schema.context {
+            self.process_action_context(ctx, namespace, &mut context_entities)?;
```

Comment:
If this method takes empty HashSet why it can't create and return it?

Reply:
refactored here: [840a0f6](https://github.com/JanssenProject/jans/pull/10387/commits/840a0f6334d717ceac542a20f63eea63b0d052c2)